src/role_scout.js

```diff
@@ -13,153 +13,226 @@ roles.scout.settings = {
+    if (scoutSeen) {
+      if (Game.creeps[scoutSeen]) {
+        continue;
       }
+      delete Memory.rooms[room].scoutSeen;
+    }
+    if (Memory.rooms[room].lastSeen < oldestAge) {
+      oldestRoom = room;
+      oldestAge = Memory.rooms[room].lastSeen;
+    }
+  }
+  if (oldestRoom) {
+    creep.memory.search.target = oldestRoom;
+    return true;
+  }
+  return false;
+};
 
-    } catch (e) {
-      if (e !== null) {
-        creep.log(`search: ${targetPosObject} ${e} ${e.stack}`);
-      } else {
-        creep.log(`search: ${targetPosObject} ${e}`);
-      }
-      // creep.memory.search.seen.push(creep.memory.search.target);
-      // // TODO extract to a method
-      // if (!setNewTarget(creep)) {
-      //   creep.memory.search.levels.push([]);
-      //   for (let room of creep.memory.search.levels[creep.memory.search.level]) {
-      //     let rooms = Game.map.describeExits(room);
-      //     for (let direction in rooms) {
-      //       let roomNext = rooms[direction];
-      //       if (haveNotSeen(creep, roomNext)) {
-      //         creep.memory.search.levels[creep.memory.search.level + 1].push(roomNext);
-      //         creep.memory.search.target = roomNext;
-      //       }
-      //     }
-      //   }
-      //   creep.memory.search.level++;
-      // }
-      return false;
+roles.scout.goFurther = function(creep) {
+  creep.memory.search.levels[creep.memory.search.level + 1] = [];
+  let target = roles.scout.lookAround(creep, creep.memory.search.level + 1, true);
+  if (target) {
+    if (!Memory.rooms[target]) {
+      Memory.rooms[target] = {};
     }
+    Memory.rooms[target].scoutSeen = creep.name;
+  }
+
+  if (creep.memory.search.levels[creep.memory.search.level + 1].length === 0) {
+    return false;
+  } // debug line
+  creep.memory.search.level++;
+};
+
+roles.scout.switchTarget = function(creep) {
+  if (creep.memory.scoutSkip) {
+    creep.memory.skip.push(creep.memory.search.target);
+    delete creep.memory.scoutSkip;
+  } else if (creep.room.name === creep.memory.search.target) {
+    creep.memory.search.seen.push(creep.room.name);
+  }
+  if (!roles.scout.lookAround(creep, 0, false, true) &&
+  !roles.scout.setNewTarget(creep)) {
+    roles.scout.goFurther(creep);
+  }
+};
 
-    if (creep.memory.last && creep.memory.last.pos3 && creep.pos.roomName !== creep.memory.last.pos3.roomName) {
-      creep.moveTo(25, 25);
-      return true;
+roles.scout.move = function(creep) {
+  let costMatrix = function(roomName) {
+    if (!Memory.rooms[roomName] || (Memory.rooms[roomName].lastSeen < (Game.time - 5))) {
+      return creep.room.getCostMatrixCallback(creep.targetPosObject, true, false, true);
     }
+    return false;
+  };
 
-    if (creep.isStuck()) {
-      creep.moveRandom();
-      creep.say('ImStuck', true);
-      creep.log('Scout Stuck, Randomly Moving: ' + JSON.stringify(creep.memory.last) + ' ' + JSON.stringify(creep.isStuck()));
-      return true;
+  let search = PathFinder.search(
```

Comment:
what about caching path in memory? at least for several ticks, like `moveTo`. Or maybe encapsulate search inside `moveToMy` somehow, and make cache layer there later.

Reply:
as pathing layout don't use a ratio 2:1 as plain/swamp we can't use moveToMy as it is actually.
but we can for sure memorize path by creep.
already thought about memorize rooms deep, which is memorized in scout, into base room but as it is working actually, i thought about do it in next PR.

Reply:
I'm also considering caching path in memory. Currently, when scout move from room A to room B, room A may become invisible, and `PathFinder` gives different search result because  `getCostMatrixCallback` returns a different `costMatrix` for room A. As a result, scout frequently gets stuck on borders. There are two solutions:
1. Cache path in memory, thus scout won't be influenced by different path search results.
2. Cache another version of `costMatrix` in memory for scout.
I prefer the first solution because it's not hard to be implented, and can benefit all roles that use `pathFinder`.

Reply:
^^ That is a really good finding. Had that issue a couple of days ago, couldn't find the reason.

I like the 2., too. Caching the complete path would result in less CPU usage.

Reply:
the issue is definitly fixed by  `if (!Memory.rooms[roomName] || (Memory.rooms[roomName].lastSeen < (Game.time - 5))) {` before use costatrix.
the issue caching path/Cm is that depending the number of rooms / scouts by base rooms, it will surely quick go to a leack of memory. already seen problems with the size of search.levels while using multiple scouts.